pkg/opencontainer/containers.go

```diff
@@ -0,0 +1,294 @@
+		return nil, err
+	}
+	l.spec = &spec
+	return l.spec, nil
+}
+
+// ID implements the launch.Launchable interface. It returns the name of this launchable.
+func (l *Launchable) ID() string {
+	return l.ID_
+}
+
+// The version of the artifact is currently derived from the location, using
+// the naming scheme <the-app>_<unique-version-string>.tar.gz
+func (hl *Launchable) Version() string {
+	fileName := filepath.Base(hl.Location)
+	return fileName[:len(fileName)-len(".tar.gz")]
+}
+
+func (*Launchable) Type() string {
+	return "opencontainer"
+}
+
+// Fetcher returns a uri.Fetcher that is capable of fetching the launcahble's files.
+func (l *Launchable) Fetcher() uri.Fetcher {
+	return uri.DefaultFetcher
+}
+
+// InstallDir is the directory where this launchable should be installed.
+func (l *Launchable) InstallDir() string {
+	launchableName := l.Version()
+	return filepath.Join(l.RootDir, "installs", launchableName)
+}
+
+// Executables gets a list of the runit services that will be built for this launchable.
+func (l *Launchable) Executables(serviceBuilder *runit.ServiceBuilder) ([]launch.Executable, error) {
+	if !l.Installed() {
+		return []launch.Executable{}, util.Errorf("%s is not installed", l.ID_)
+	}
+
+	uid, gid, err := user.IDs(l.RunAs)
+	if err != nil {
+		return nil, util.Errorf("%s: unknown runas user: %s", l.ID_, l.RunAs)
+	}
+	lspec, err := l.getSpec()
+	if err != nil {
+		return nil, util.Errorf("%s: loading container specification: %s", l.ID_, err)
+	}
+	expectedPlatform := Platform{
+		OS:   "linux",
+		Arch: "amd64",
+	}
+	if lspec.Platform != expectedPlatform {
+		return nil, util.Errorf(
+			"%s: unsupported container platform: %#v expected %#v",
+			l.ID_,
+			lspec.Platform,
+			expectedPlatform,
+		)
+	}
+	if filepath.Base(lspec.Root.Path) != lspec.Root.Path {
+		return nil, util.Errorf("%s: invalid container root: %s", l.ID_, lspec.Root.Path)
+	}
+	luser := lspec.Process.User
+	if uid != int(luser.UID) || gid != int(luser.GID) {
+		return nil, util.Errorf("%s: cannot execute as %s(%d:%d): container expects %d:%d",
+			l.ID_, l.RunAs, uid, gid, luser.UID, luser.GID)
+	}
+
+	serviceName := l.ID_ + "__container"
+	return []launch.Executable{{
+		Service: runit.Service{
+			Path: filepath.Join(serviceBuilder.RunitRoot, serviceName),
+			Name: serviceName,
+		},
+		Exec: append(
+			[]string{l.P2Exec},
+			p2exec.P2ExecArgs{
+				NoLimits: true,
+				WorkDir:  l.InstallDir(),
+				Command:  []string{*RuncPath, "start"},
+			}.CommandLine()...,
+		),
+	}}, nil
+}
+
+// Installed returns true if this launchable is already installed.
+func (l *Launchable) Installed() bool {
+	installDir := l.InstallDir()
+	_, err := os.Stat(installDir)
+	return err == nil
+}
+
+// Install ...
+func (l *Launchable) Install() (returnedError error) {
+	if l.Installed() {
+		return nil
+	}
+
+	data, err := uri.DefaultFetcher.Open(l.Location)
+	if err != nil {
+		return err
+	}
+	defer data.Close()
+	defer func() {
+		if returnedError != nil {
```

Comment:
Got it, I didn't realize calling `return err` would implicitly assign the error to the variable named by the return clause.